com.avaloq.tools.ddk.xtext.builder/src/com/avaloq/tools/ddk/xtext/builder/CachingStateBasedContainerManager.java

```diff
@@ -126,11 +127,26 @@ public void descriptionsChanged(final Event event) {
       return Collections.emptyList();
     }
     List<String> handles = getState(resourceDescriptions).getVisibleContainerHandles(root);
+    return getContainersForHandlesAndResource(handles, desc, resourceDescriptions);
+  }
+
+  /**
+   * Returns the containers for the given handles, where one of the containers will also include {@code desc} when appropriate.
+   * 
+   * @param handles
+   *          handles to get containers for, must not be {@code null}
+   * @param desc
+   *          description to add, must not be {@code null}
+   * @param resourceDescriptions
+   *          resource descriptions, must not be {@code null}
+   * @return list of containers, never {@code null}
+   */
+  protected List<IContainer> getContainersForHandlesAndResource(final List<String> handles, final IResourceDescription desc, final IResourceDescriptions resourceDescriptions) {
     List<IContainer> result = getVisibleContainers(handles, resourceDescriptions);
     if (!result.isEmpty()) {
       URI descURI = desc.getURI();
-      for (IContainer container : result) {
-        if (container.getResourceDescription(descURI) != null) {
+      for (int i = 0; i < result.size(); i++) {
+        if (result.get(i).getResourceDescription(descURI) != null) {

```

Comment:
You now rely on result being an array list.  Should we document it somewhere? getVisibleContainers can be overridden and implementers must be aware of this detail.

Reply:
We have one IContainer for each handle - do i remember correct? If yes, this list is always small enough not to care about this

Reply:
We could create a follow-up and check if the list implements RandomAccess.